### DISCUSSION TITLE
Is there an official NuGet single-string package identifier?

### DISCUSSION THREAD

**Discussion by HID-jthierry** [NuGet/Home#11853] [Q&A] [answered]
Hi,

For a dependency management system, I'm looking to see if there is an official single-string format identifying NuGet packages, which combines package name and version. I couldn't find any information on this, the official documentation always separating the package name from the version.

I am thinking for example of what is done for Maven, npm or pip:
Maven: `groupId:artifactId:version`
npm: `<name>@<version>`
pip: `name==version`

Thanks

**Comment by nkolev92** [comment, marked as answer]
In packages.config world, NuGet uses a 1 level packages folder, and it was `id.version`, but that that does not 100% roundtrip, since you can't separate between something like: 

`package`.`1.0.0.1` and `package1`.`0.0.1`

NuGet doesn't really use a single identifier for simplicity sake. Our declaration language for packages has 2 separate values. 

Anything that's not a valid package id char or a version char can work as a separator though. 

How would you use that separator? 

In our build intermediaries, we sometimes use `/`

**Reply by HID-jthierry** [reply]
Thanks nkolev92

It totally makes sense to use a character that's not a valid package id char.

This automatically brings the question what is a valid package id char?. As you are using `/` in your build intermediaries, I suppose that `/` is NOT a valid package id char.

I also understand that `.` is a valid package id char as Microsoft and NuGet themselves use `.` in their package id prefixes (*).

Now how can I know if non-alphanumeric characters such as `[(-+_!%,;*:@?^<>~#)]` are valid package id chars or not? Is this documented somewhere? Or is there a piece of code I can read to understand how package ids are parsed?

(*) https://docs.microsoft.com/en-us/nuget/nuget-org/id-prefix-reservation

**Reply by nkolev92** [reply]
Try: 

https://github.com/NuGet/NuGet.Client/blob/d692f4ce5ec417f906da0442737b32c158bb2bb9/src/NuGet.Core/NuGet.Packaging/PackageCreation/Utility/PackageIdValidator.cs#L18
